pkg/controllers/gitrepository/controller.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	err = writeRepoContents(repo, tempDir)
+	if err != nil {
+		return err
+	}
+
+	tree, err := clonedRepo.Worktree()
+	if err != nil {
+		return fmt.Errorf("getting git worktree: %w", err)
+	}
+
+	err = tree.AddGlob("*")
+	if err != nil {
+		return fmt.Errorf("adding git files: %w", err)
+	}
+
+	status, err := tree.Status()
+	if err != nil {
+		return fmt.Errorf("getting git status: %w", err)
+	}
+
+	if status.IsClean() {
+		h, _ := clonedRepo.Head()
+		repo.Status.LatestCommit.Hash = h.Hash().String()
+		return nil
+	}
+
+	commit, err := tree.Commit(fmt.Sprintf("updated from %s", repo.Spec.Source.Path), &git.CommitOptions{
+		All:               true,
+		AllowEmptyCommits: false,
+		Author: &object.Signature{
+			Name:  gitCommitAuthorName,
+			Email: gitCommitAuthorEmail,
+			When:  time.Now(),
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("committing git files: %w", err)
+	}
+
+	auth, err := getBasicAuth(repo.Name, repo.Namespace)
+	if err != nil {
+		return fmt.Errorf("getting basic auth: %w", err)
+	}
+	err = clonedRepo.Push(&git.PushOptions{
+		Auth: &auth,
+	})
+	if err != nil {
+		return fmt.Errorf("pushing to git: %w", err)
+	}
+
+	repo.Status.LatestCommit.Hash = commit.String()
+
+	return nil
+}
+
+func reconcileRepo(giteaClient GiteaClient, repo *v1alpha1.GitRepository) (*gitea.Repository, error) {
+	resp, repoResp, err := giteaClient.GetRepo(getOrganizationName(*repo), getRepositoryName(*repo))
+	if err != nil {
+		if repoResp.StatusCode == 404 {
+			createResp, _, CErr := giteaClient.CreateOrgRepo(getOrganizationName(*repo), gitea.CreateRepoOption{
+				Name:        getRepositoryName(*repo),
+				Description: fmt.Sprintf("created by Git Repository controller for %s in %s namespace", repo.Name, repo.Namespace),
+				// we should reconsider this when targeting non-local clusters.
+				Private:       false,
+				DefaultBranch: DefaultBranchName,
+				AutoInit:      true,
+			})
+			if CErr != nil {
+				return &gitea.Repository{}, fmt.Errorf("failed to create git repository: %w", CErr)
+			}
+			repo.Status.ExternalGitRepositoryUrl = createResp.CloneURL
+			return createResp, nil
+		}
+	}
+	return resp, nil
+}
+
+func (r *RepositoryReconciler) SetupWithManager(mgr ctrl.Manager, notifyChan chan event.GenericEvent) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.GitRepository{}).
+		Watches(&source.Channel{Source: notifyChan}, &handler.EnqueueRequestForObject{}).
+		Complete(r)
+}
+
+func (r *RepositoryReconciler) shouldProcess(repo v1alpha1.GitRepository) bool {
+	if repo.Spec.Source.Type == "local" && !filepath.IsAbs(repo.Spec.Source.Path) {
+		return false
+	}
+	// embedded fs does not change
+	if repo.Spec.Source.Type == "embedded" && repo.Status.Synced {
+		return false
+	}
+	return true
+}
+
+func getEmbedded(name string) ([][]byte, error) {
+	switch name {
+	case "argocd":
+		return localbuild.GetRawInstallResources()
+	case "nginx":
+		return localbuild.RawNginxInstallResources()
+	default:
```

Comment:
ah, now I see your point! this breaks since in the previous commit I removed these functions. maybe we put the functions back where needed?

Reply:
Also, probably makes more sense to decouple resource creation from the actual git repository controller code and place it all here where we reconcile other resources as well

https://github.com/cnoe-io/idpbuilder/blob/11a9914536cbd8ee8da9f5d79e8924f1e595ec0c/pkg/controllers/localbuild/controller.go#L280-L286C2

Reply:
I agree with this, let's have one place in the sources for the creation of resources no matter what their install type

Reply:
Yep that makes sense.